transformer_engine/pytorch/transformer.py

```diff
@@ -219,6 +228,243 @@ def forward(
+    """
+
+    def __init__(
+        self,
+        num_attention_heads: int,
+        kv_channels: int,
+        attention_dropout: float = 0.0,
+        layer_number: Optional[int] = None,
+        apply_query_key_layer_scaling: bool = True,
+        attention_softmax_in_fp32: bool = True,
+        attn_mask_type: str = "causal",
+        sequence_parallel: bool = False,
+        tp_size: int = 1,
+        get_rng_state_tracker: Optional[Callable] = None,
+    ) -> None:
+        super().__init__()
+
+        self.use_flash_attention = int(os.getenv("NVTE_FLASH_ATTN", "1"))
+        if self.use_flash_attention:
+            attention = FlashAttention
+        else:
+            attention = UnfusedDotProductAttention
+
+        self.attention = attention(
+            num_attention_heads,
+            kv_channels,
+            attention_dropout,
+            layer_number=layer_number,
+            apply_query_key_layer_scaling=apply_query_key_layer_scaling,
+            attention_softmax_in_fp32=attention_softmax_in_fp32,
+            tp_size=tp_size,
+            get_rng_state_tracker=get_rng_state_tracker,
+            attn_mask_type=attn_mask_type,
+            sequence_parallel=sequence_parallel,
+        )
+
+    def _checkpointed_attention_forward(
+        self,
+        *forward_args: Tuple[torch.Tensor, ...],
+    ) -> torch.Tensor:
+        """Forward method with activation checkpointing."""
+
+        def custom_forward(*inputs):
+            return self.attention(*inputs)
+
+        hidden_states = checkpoint(
+            custom_forward,
+            False,
+            self.get_rng_state_tracker,
+            self.tp_group,
+            *forward_args,
+        )
+
+        return hidden_states
+
+
+    def forward(
+        self,
+        query_layer: torch.Tensor,
+        key_layer: torch.Tensor,
+        value_layer: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        checkpoint_core_attention: bool = False,
+    ) -> torch.Tensor:
+        """
+        Dot Product Attention Layer.
+
+        Parameters
+        ----------
+        query_layer : torch.Tensor
+                     Query tensor.
+        key_layer : torch.Tensor
+                   Key tensor.
+        value_layer : torch.Tensor
+                     Value tensor.
+        attention_mask : torch.Tensor
+                        Boolean tensor used to mask out softmax input when not using flash-attn.
+        checkpoint_core_attention : bool, default = `True`
+                                   If true, forward activations for attention are recomputed
+                                   during the backward pass in order to save memory that would
+                                   otherwise be occupied to store the forward activations until
+                                   backprop.
+        """
+
+        if self.use_flash_attention:
+            assert attention_mask is None, "flash-attn does not require attention_mask."
```

Comment:
That should not be an assert - if the attention mask is here, you should just not use flash attention.

Or better yet - have mutually exclusive options for the attention_mask being boolean or a tensor of cumulative sequence lengths that flash attention needs (the unpadded case basically).